zkevm-circuits/src/mpt_circuit.rs

```diff
@@ -0,0 +1,719 @@
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap();
+
+        let memory_columns = (0..5).map(|_| meta.advice_column()).collect::<Vec<_>>();
+
+        let mut state_machine = StateMachineConfig::construct(meta);
+        let mut rlp_item = MainRLPGadget::default();
+
+        let mut memory = Memory::new(memory_columns);
+        memory.allocate(meta, key_memory(false));
+        memory.allocate(meta, key_memory(true));
+        memory.allocate(meta, parent_memory(false));
+        memory.allocate(meta, parent_memory(true));
+        memory.allocate(meta, main_memory());
+
+        let mut ctx = MPTContext {
+            mpt_table,
+            rlp_item: rlp_item.clone(),
+            challenges: challenges.clone(),
+            r: challenges.keccak_input(),
+            memory: memory.clone(),
+        };
+
+        let rlp_cm = CellManager::new(
+            meta,
+            // Type, #cols, phase, permutable
+            vec![
+                (MptCellType::StoragePhase1, 50, 0, false),
+                (MptCellType::StoragePhase2, 5, 0, false),
+                (MptCellType::Lookup(Table::Fixed), 3, 0, false),
+            ],
+            0,
+            1,
+        );
+        let state_cm = CellManager::new(
+            meta,
+            // Type, #cols, phase, permutable
+            vec![
+                (MptCellType::StoragePhase1, 20, 0, false),
+                (MptCellType::StoragePhase2, 5, 0, false),
+                (MptCellType::LookupByte, 4, 0, false),
+                (MptCellType::Lookup(Table::Fixed), 2, 0, false),
+                (MptCellType::Lookup(Table::Keccak), 1, 0, false),
+            ],
+            0,
+            50,
+        );
+
+        let mut cb = MPTConstraintBuilder::new(5, Some(challenges.clone()), None);
+        meta.create_gate("MPT", |meta| {
+            circuit!([meta, cb], {
+                // Populate lookup tables
+                require!(@KECCAK => <KeccakTable as LookupTable<F>>::advice_columns(&keccak_table).iter().map(|table| a!(table)).collect());
+                require!(@FIXED => fixed_table.iter().map(|table| f!(table)).collect());
+
+                ifx!{f!(q_enable) => {
+                    // RLP item decoding unit
+                    cb.base.set_cell_manager(rlp_cm.clone());
+                    cb.base.push_region(MPTRegion::RLP as usize);
+                    rlp_item = MainRLPGadget::construct(&mut cb, &ctx.r);
+                    cb.base.pop_region();
+                    ctx.rlp_item = rlp_item.clone();
+
+                    // Main MPT circuit
+                    // State machine
+                    cb.base.set_cell_manager(state_cm.clone());
+                    ifx! {f!(q_first) + f!(q_last) => {
+                        require!(a!(state_machine.is_start) => true);
+                    }};
+                    // Main state machine
+                    matchx! {
+                        a!(state_machine.is_start) => {
```

Comment:
@Brechtpd I think there are no constraints currently about which nodes can follow which nodes, but after some thought I think these are indeed not necessary as (in `account_leaf.rs`)
```
require!(config.main_data.is_below_account => false);
```
guarantees there are no two or more account leaves. Another possible attack would be to have two storage proofs one after another, but the probability is again negligible as the second storage trie would need to hash to 0 as (the first) `storage_leaf` stores the default values (`parent_data.rlc = 0`). So I think it's all good :)

Reply:
Yes I also believe it should be fine but could miss some case. I actually thought `is_below_account` was set to false again after the storage leaf, but looks like that isn't the case so seem like the hash being `0` is indeed the only thing stopping it from happening. Perhaps it can't hurt to also update `is_below_account` though. :)